tasks/human_eval_025.rs

```diff
@@ -5,14 +5,377 @@ HumanEval/25
 /*
 ### VERUS BEGIN
 */
+use vstd::arithmetic::div_mod::*;
+use vstd::arithmetic::mul::*;
+use vstd::assert_by_contradiction;
+use vstd::calc;
 use vstd::prelude::*;
 
 verus! {
 
-// TODO: Put your solution (the specification, implementation, and proof) to the task here
+pub closed spec fn is_prime(n: nat) -> bool {
+    forall|i: nat| 1 < i < n ==> #[trigger] (n % i) != 0
+}
+
+// canonical definition of prime factoriztion
+pub closed spec fn is_prime_factorization(n: nat, factorization: Seq<nat>) -> bool {
+    // all factors are prime
+    &&& forall|i: nat|
+        0 <= i < factorization.len() ==> #[trigger] is_prime(
+            factorization[i as int] as nat,
+        )
+    // product of factors is n
+    &&& factorization.fold_right(|x: nat, acc: nat| (acc * x as nat), 1nat)
+        == n
+    // factors are listed in ascending order
+    &&& forall|i: nat, j: nat|
+        (#[trigger] (1 + i - i + j - j) < i <= j < factorization.len()) ==> (factorization[i as int]
```

Comment:
The expression `(1 + i - i + j - j)` instead of `1` is a bit surprising.  Could you instead trigger off of `factorization[i as int]` and `factorization[j as int]`?

Reply:
The problem with this, is it is really hard to get this quantifier to trigger at the end of the factorize function. I think the reason for this, is because we are now replacing factorization with `factorization@.map(|_idx, j: u8| j as nat)` and so triggering on `factorization[i as int]` and `factorization[j as int]` won't cause a trigger

This was the cleanest solution I could find, even though it does look a little silly

Reply:
In general, we should be suspicious if we find ourselves contorting the trusted spec to facilitate a better proof.  In this case, I think you can use the cleaner definition in the trusted `is_prime_factorization`, stick with the "messy"-but-convenient trigger strategy you're using in your untrusted loop invariant, and then just make sure the necessary trigger fires to make the proof go through.  In this case, adding:
```
    proof {
        let f_seq = factorization@.map(|_idx, j: u8| j as nat);
        assert forall|i: nat, j: nat| (1 < i <= j < f_seq.len()) implies (f_seq[i as int] <= f_seq[j as int]) by {
            assert(1 + i - i + j - j > 0);
        }
    }
```
at the end of `factorize` [seems to work](https://play.verus-lang.org/?version=stable&mode=basic&edition=2021&gist=64993a184502c1229d993fec53658601).